### PR TITLE
Fix IncrementalDelaunayTriangulator to ensure triangulation boundary is convex

### PR DESCRIPTION
**Pull request by dr-jts**
Port of https://github.com/locationtech/jts/pull/1004

This adds logic to `IncrementalDelaunayTriangulator` to ensure that the triangulation boundary is always convex (as per specification). This is required because the "frame" created to contain the triangulation construction is not infinitely distant from the input points. This has been causing robustness failures in cases where a site lies very close to the convex hull of the input.

This is a robust enhancement that replaces the frame size heuristic change in #728.

Also fixes issues reported for ConcaveHull ( #719, #946)

### CODE REVIEW

File: src/triangulate/IncrementalDelaunayTriangulator.cpp [hunk 1]
```diff
@@ -82,25 +91,76 @@ IncrementalDelaunayTriangulator::insertSite(const Vertex& v)
...
+    }
+    return false;
+}
+
+bool 
+IncrementalDelaunayTriangulator::isConcaveAtOrigin(const QuadEdge& e) 
+{
+    Coordinate p = e.orig().getCoordinate();
+    Coordinate pp = e.oPrev().dest().getCoordinate();
+    Coordinate pn = e.oNext().dest().getCoordinate();
```
Comment (dbaston) [inline src/triangulate/IncrementalDelaunayTriangulator.cpp:150]: Use `const Coordinate&` here

File: src/triangulate/IncrementalDelaunayTriangulator.cpp [hunk 2]
```diff
@@ -82,25 +91,76 @@ IncrementalDelaunayTriangulator::insertSite(const Vertex& v)
...
+    bool isConcave = Orientation::COUNTERCLOCKWISE == Orientation::index(pp, pn, p);
+    return isConcave;
+}
+
+bool 
+IncrementalDelaunayTriangulator::isBetweenFrameAndInserted(const QuadEdge& e, const Vertex& vInsert) 
+{
+    const Vertex v1 = e.oNext().dest();
+    const Vertex v2 = e.oPrev().dest();
```
Comment (dbaston) [inline src/triangulate/IncrementalDelaunayTriangulator.cpp:159]: `const Vertex&`

File: include/geos/triangulate/IncrementalDelaunayTriangulator.h [hunk 3]
```diff
@@ -77,6 +91,14 @@ class GEOS_DLL IncrementalDelaunayTriangulator {
      * @return a quadedge containing the inserted vertex
      */
     quadedge::QuadEdge& insertSite(const quadedge::Vertex& v);
+
+private:
+
+    bool isConcaveBoundary(const quadedge::QuadEdge& e);
```
Comment (dbaston) [inline include/geos/triangulate/IncrementalDelaunayTriangulator.h:97]: methods can be `const`

**Comment by dr-jts** [comment]
`const` comments fixed in https://github.com/libgeos/geos/commit/1033161eca3684e0d9479daeafd35a4e0fb1800e.

Thanks.

**Pull request merged**